### PR TITLE
Hw dao6

### PR DESCRIPTION
**Pull request by engineer027**

### CODE REVIEW

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        String query = "INSERT INTO cars (model_car, manufacturer_id) VALUES (?, ?);";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query, Statement
+                        .RETURN_GENERATED_KEYS)) {
+            statement.setString(1, car.getModel());
+            statement.setLong(2, car.getManufacturer().getId());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                car.setId(resultSet.getLong(1));
```
Comment (RuslanIbIsh) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: в задании предлагают использовать getObject(name, Long.class)

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                + "manufacturer_id = ? WHERE id_car = ? "
+                + "AND delete_car = FALSE";
+        removeDriverFromCar(car);
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setString(1, car.getModel());
+            statement.setLong(2, car.getManufacturer().getId());
+            statement.setLong(3, car.getId());
+            statement.executeUpdate();
+            for (Driver driver : car.getDrivers()) {
```
Comment (RuslanIbIsh) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: в задании или в чеклисте есть условие, что нужно сначала стереть всех водителей из cars_drivers, а потом записать

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            }
+            return car;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could update car " + car, e);
+        }
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        String query = "UPDATE cars SET delete_car = ? "
```
Comment (RuslanIbIsh) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
        String query = "UPDATE cars SET delete_car = true "
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        String query = "SELECT cars.id_car, cars.model_car, cars.manufacturer_id, "
+                + "manufacturer.name_manufacturer, manufacturer.country_manufacturer "
+                + "FROM cars "
+                + "JOIN cars_drivers ON cars.id_car = cars_drivers.id_car "
+                + "JOIN manufacturer ON cars.manufacturer_id = manufacturer.id_manufacturer "
+                + "WHERE cars_drivers.id_driver = ? "
+                + "AND delete_car = FALSE";
```
Comment (RuslanIbIsh) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: в условиях к заданию есть упоминание что нужно проверять и drivers and cars - deleted or not

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+            statement.setLong(1, car.getId());
+            statement.setLong(2, driver.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not add driver "
+                    + driver + "to car " + car, e);
+        }
+    }
+
+    private void removeDriverFromCar(Car car) {
```
Comment (RuslanIbIsh) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: у тебя в методе удаляются несколько драйверов - а метод имеет название про одного драйвера

**Review by RuslanIbIsh — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java (hunk 2 above)
Comment (engineer027) [reply src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: я це і роблю на 87 ст за допомогою removeDriverFromCar(car)

**Review by RuslanIbIsh — APPROVED** [review]

File: src/main/resources/init_db.sql [hunk 6]
```diff
@@ -23,3 +23,67 @@ ALTER TABLE `taxi_service`.`manufacturer`
     CHANGE COLUMN `nameManufacturer` `name_manufacturer` VARCHAR(225) NOT NULL ,
     CHANGE COLUMN `countryManufacturer` `country_manufacturer` VARCHAR(225) NOT NULL ,
     CHANGE COLUMN `deleteManufacturer` `delete_manufacturer` TINYINT NOT NULL DEFAULT '0' ;
+
+CREATE TABLE `taxi_service`.`drivers` (
+                                          `id_driver` BIGINT NOT NULL,
+                                          `name_driver` VARCHAR(225) NOT NULL,
+                                          `licence_number_driver` VARCHAR(225) NOT NULL,
+                                          `delete_driver` TINYINT NOT NULL DEFAULT 0,
```
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql]:
```suggestion
                                          `id` BIGINT NOT NULL,
                                          `name` VARCHAR(225) NOT NULL,
                                          `licence` VARCHAR(225) NOT NULL,
                                          `delete` TINYINT NOT NULL DEFAULT 0,
```
same for `manufacturer` and `car` tables - don't repeat the table name in columns

File: src/main/resources/init_db.sql [hunk 7]
```diff
@@ -23,3 +23,67 @@ ALTER TABLE `taxi_service`.`manufacturer`
...
+                                          PRIMARY KEY (`id_driver`),
+                                          UNIQUE INDEX `name_driver_UNIQUE` (`name_driver` ASC) VISIBLE,
+                                          UNIQUE INDEX `licence_number_driver_UNIQUE` (`licence_number_driver` ASC) VISIBLE)
+    ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8
+COLLATE = utf8_bin;
+
+ALTER TABLE `taxi_service`.`drivers`
+    CHANGE COLUMN `id_driver` `id_driver` BIGINT NOT NULL AUTO_INCREMENT ;
```
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql]: let's put here the create table statements that are correct from the start

File: src/main/resources/init_db.sql [hunk 8]
```diff
@@ -23,3 +23,67 @@ ALTER TABLE `taxi_service`.`manufacturer`
...
+
+CREATE TABLE `taxi_service`.`cars` (
+                                       `id_car` BIGINT NOT NULL AUTO_INCREMENT,
+                                       `model_car` VARCHAR(225) NOT NULL,
+                                       `manufacturer_id` BIGINT NOT NULL,
+                                       `delete_car` TINYINT NOT NULL DEFAULT 0,
+                                       PRIMARY KEY (`id_car`),
+                                       UNIQUE INDEX `carscol_UNIQUE` (`model_car` ASC) VISIBLE,
```
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql]: why is model unique?

File: src/main/resources/init_db.sql [hunk 9]
```diff
@@ -23,3 +23,67 @@ ALTER TABLE `taxi_service`.`manufacturer`
...
+INSERT INTO `taxi_service`.`drivers` (`name_driver`, `licence_number_driver`) VALUES ('Alice', '666');
+
+INSERT INTO `taxi_service`.`cars` (`model_car`, `manufacturer_id`) VALUES ('Fafia', '1');
+INSERT INTO `taxi_service`.`cars` (`model_car`, `manufacturer_id`) VALUES ('Lanos', '2');
+
+ALTER TABLE `taxi_service`.`cars`
+    CHANGE COLUMN `delete_car` `delete_car` TINYINT NOT NULL DEFAULT '0' ;
+
+ALTER TABLE `taxi_service`.`cars_drivers`
+    RENAME TO  `taxi_service`.`cars_drivers` ;
```
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql]: remove this and put to this file correct create statements

File: src/main/java/mate/academy/dao/jdbc/DriverDaoJdbc.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    @Override
+    public Optional<Driver> get(Long id) {
+        String query = "SELECT * FROM drivers where id_driver = ?"
+                + " AND delete_driver = FALSE";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(createDriver(resultSet)); // cndjhbnb
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/DriverDaoJdbc.java]: remove the comment

File: src/main/java/mate/academy/dao/jdbc/DriverDaoJdbc.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+            statement.executeUpdate();
+            return driver;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could update driver" + driver, e);
+        }
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        String query = "UPDATE drivers SET delete_driver = ? "
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/DriverDaoJdbc.java]:
```suggestion
        String query = "UPDATE drivers SET delete_driver = TRUE "
```

File: src/main/java/mate/academy/dao/jdbc/DriverDaoJdbc.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+            statement.setLong(2, id);
+            int delete = statement.executeUpdate();
+            return delete > 0;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not delete driver by id" + id, e);
+        }
+    }
+
+    private Driver createDriver(ResultSet resultSet) throws SQLException {
+        Long idDriver = resultSet.getObject("id_driver", long.class);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/DriverDaoJdbc.java]:
```suggestion
        Long idDriver = resultSet.getObject("id_driver", Long.class);
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+import mate.academy.exception.DataProcessingException;
+import mate.academy.lib.Dao;
+import mate.academy.model.Car;
+import mate.academy.model.Driver;
+import mate.academy.model.Manufacturer;
+import mate.academy.util.ConnectionUtil;
+
+@Dao
+public class CarDaoJdbc implements CarDao {
+
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: redundant empty line

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+        }
+        return car;
+    }
+
+    @Override
+    public Optional<Car> get(Long id) {
+        String query = "SELECT cars.id_car, cars.model_car, cars.manufacturer_id, "
+                + "manufacturer.name_manufacturer, manufacturer.country_manufacturer "
+                + "FROM cars "
+                + "JOIN manufacturer ON cars.manufacturer_id = manufacturer.id_manufacturer "
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
                + "FROM cars c "
                + "JOIN manufacturer m ON c.manufacturer_id = m.id_manufacturer "
```
use aliases

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(createCar(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not get car by id " + id, e);
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
        Car car = null;
        try (Connection connection = ConnectionUtil.getConnection();
                PreparedStatement statement = connection.prepareStatement(query)) {
            statement.setLong(1, id);
            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                car = createCar(resultSet);
            }
        } catch (SQLException e) {
            throw new DataProcessingException("Could not get car by id " + id, e);
        }
        if (car != null) {
            // get drivers of the car in a separate method that opens a separate connection
        }
        return Optional.ofNullable(car);
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+        List<Car> cars = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                cars.add(createCar(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not get all cars ", e);
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: same for this method

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+                + "AND delete_driver = FALSE "
+                + "AND delete_car = FALSE";
+        List<Car> cars = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, driverId);
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                cars.add(createCar(resultSet));
+            }
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java:139]: same for this method

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        String query = "SELECT cars.id_car, cars.model_car, cars.manufacturer_id, "
+                + "manufacturer.name_manufacturer, manufacturer.country_manufacturer "
+                + "FROM cars "
+                + "JOIN cars_drivers ON cars.id_car = cars_drivers.id_car "
+                + "JOIN manufacturer ON cars.manufacturer_id = manufacturer.id_manufacturer "
+                + "WHERE cars_drivers.id_driver = ? "
+                + "AND delete_driver = FALSE "
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: should we join `drivers` to have access to `delete_driver` column?

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+            statement.setLong(2, car.getManufacturer().getId());
+            statement.setLong(3, car.getId());
+            statement.executeUpdate();
+            for (Driver driver : car.getDrivers()) {
+                addDriverToCar(driver, car);
+            }
+            return car;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could update car " + car, e);
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
        try (Connection connection = ConnectionUtil.getConnection();
                PreparedStatement statement = connection.prepareStatement(query)) {
            statement.setString(1, car.getModel());
            statement.setLong(2, car.getManufacturer().getId());
            statement.setLong(3, car.getId());
            statement.executeUpdate();
        } catch (SQLException e) {
            throw new DataProcessingException("Could update car " + car, e);
        }
        deleteAllDriversFromCar(car);
        addDriversToCar(car);
        return car;
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+            throw new DataProcessingException("Could not get all Cars by Driver id " + driverId, e);
+        }
+        return cars;
+    }
+
+    private void addDriverToCar(Driver driver, Car car) {
+        String query = "INSERT INTO cars_drivers (id_car, id_driver) VALUES (?, ?);";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, car.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
    private void addDriverToCar(Car car) {
        String query = "INSERT INTO cars_drivers (id_car, id_driver) VALUES (?, ?);";
        try (Connection connection = ConnectionUtil.getConnection();
                PreparedStatement statement = connection.prepareStatement(query)) {
                
            // for each driver: 
            statement.setLong(1, car.getId());
            ...
```
let's not open connections in a loop, we can execute statement multiple times in one connection instead

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+        Long idCar = resultSet.getObject("id_car", long.class);
+        String modelCar = resultSet.getString("model_car");
+        Long manufacturerIdCar = resultSet.getObject("manufacturer_id", long.class);
+        String nameManufacturer = resultSet.getString("name_manufacturer");
+        String countryManufacturer = resultSet.getString("country_manufacturer");
+        Manufacturer manufacturer = new Manufacturer(nameManufacturer, countryManufacturer);
+        manufacturer.setId(manufacturerIdCar);
+        Car car = new Car(modelCar, manufacturer);
+        car.setId(idCar);
+        car.setDrivers(getDriversForCar(idCar));
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
```
let's not open a new connection inside the existing one

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+        return car;
+    }
+
+    private List<Driver> getDriversForCar(Long idCar) {
+        String query = "SELECT drivers.name_driver, drivers.licence_number_driver, "
+                + "cars_drivers.id_driver "
+                + "FROM cars_drivers "
+                + "JOIN drivers ON cars_drivers.id_driver = drivers.id_driver "
+                + "WHERE cars_drivers.id_car = ?";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: use aliases and check that driver is not deleted

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/resources/init_db.sql [hunk 23]
```diff
@@ -2,24 +2,76 @@ CREATE SCHEMA `taxi_service` DEFAULT CHARACTER SET utf8;
...
-    ADD COLUMN `deleteManufacturer` TINYINT NOT NULL AFTER `countryManufacturer`;
+    ADD COLUMN `deleted` TINYINT NOT NULL AFTER `country`;
+
+CREATE TABLE `taxi_service`.`drivers` (
+                                          `id` BIGINT NOT NULL,
+                                          `name_driver` VARCHAR(225) NOT NULL,
+                                          `licence_number` VARCHAR(225) NOT NULL,
+                                          `deleted` TINYINT NOT NULL DEFAULT 0,
+                                          PRIMARY KEY (`id`),
+                                          UNIQUE INDEX `name_UNIQUE` (`name` ASC) VISIBLE,
```
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql]: let's not make the driver's name unique

File: src/main/resources/init_db.sql [hunk 24]
```diff
@@ -2,24 +2,76 @@ CREATE SCHEMA `taxi_service` DEFAULT CHARACTER SET utf8;
...
+                                          UNIQUE INDEX `licence_number_UNIQUE` (`licence_number` ASC) VISIBLE)
+    ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8
+COLLATE = utf8_bin;
+
+ALTER TABLE `taxi_service`.`drivers`
+    CHANGE COLUMN `id` `id` BIGINT NOT NULL AUTO_INCREMENT ;
```
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql]: let's put in this file the `CREATE TABLE` statements that are correct from the start

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    @Override
+    public Optional<Car> get(Long id) {
+        String query = "SELECT cars.id, cars.model, cars.manufacturer_id, "
+                + "manufacturer.name, manufacturer.country "
+                + "FROM cars c "
+                + "JOIN manufacturer m ON c.manufacturer_id = m.id "
+                + "WHERE cars.id = ? "
+                + "AND cars.deleted = FALSE";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
        String query = "SELECT c.id, c.model, c.manufacturer_id, "
                + "m.name, m.country "
                + "FROM cars c "
                + "JOIN manufacturer m ON c.manufacturer_id = m.id "
                + "WHERE c.id = ? "
                + "AND c.deleted = FALSE";
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 26]
```diff
@@ -0,0 +1,210 @@
...
+        return Optional.ofNullable(car);
+    }
+
+    @Override
+    public List<Car> getAll() {
+        String query = "SELECT cars.id, cars.model, cars.manufacturer_id, "
+                + "manufacturer.name, manufacturer.country "
+                + "FROM cars c "
+                + "JOIN manufacturer m ON c.manufacturer_id = m.id "
+                + "WHERE cars.deleted = FALSE";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: same here

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 27]
```diff
@@ -0,0 +1,210 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                cars.add(createCar(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not get all cars ", e);
+        }
+        if (cars.size() > 0) {
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: do we need this check?

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 28]
```diff
@@ -0,0 +1,210 @@
...
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                car = createCar(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not get car by id " + id, e);
+        }
+        if (car != null) {
+            getDriversForCar(car.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: let's set the drivers to car, now we just get them

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 29]
```diff
@@ -0,0 +1,210 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                cars.add(createCar(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not get all cars ", e);
+        }
+        if (cars.size() > 0) {
+            for (Car car: cars) {
+                getDriversForCar(car.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: same

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 30]
```diff
@@ -0,0 +1,210 @@
...
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        String query = "SELECT cars.id, cars.model, cars.manufacturer_id, "
+                + "manufacturer.name, manufacturer.country "
+                + "FROM cars c "
+                + "JOIN cars_drivers cd ON c.id = cd.id_car "
+                + "JOIN manufacturer m ON c.manufacturer_id = m.id "
+                + "WHERE cars_drivers.id_driver = ? "
+                + "AND cars.deleted = FALSE";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
        String query = "SELECT c.id, c.model, c.manufacturer_id, "
                + "m.name, m.country "
                + "FROM cars c "
                + "JOIN cars_drivers cd ON c.id = cd.id_car "
                + "JOIN manufacturer m ON c.manufacturer_id = m.id "
                // join drivers
                + "WHERE cd.id_driver = ? "
                + "AND c.deleted = FALSE 
                // AND the driver is not deleted";
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 31]
```diff
@@ -0,0 +1,210 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, driverId);
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                cars.add(createCar(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not get all Cars by Driver id " + driverId, e);
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java:145]: should we fill the cars with drivers?

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 32]
```diff
@@ -0,0 +1,210 @@
...
+        return cars;
+    }
+
+    private void addDriverToCar(Car car) {
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
    private void addDriversToCar(Car car) {
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 33]
```diff
@@ -0,0 +1,210 @@
...
+                statement.executeUpdate();
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not add driver "
+                    + "to car " + car, e);
+        }
+    }
+
+    private void deleteAllDriversFromCar(Car car) {
+        String query = "DELETE FROM `cars_drivers` WHERE id_car = ?";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
        String query = "DELETE FROM cars_drivers WHERE id_car = ?";
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 34]
```diff
@@ -0,0 +1,210 @@
...
+            statement.setLong(1, car.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Could not delete driver "
+                    + " from car " + car, e);
+        }
+    }
+
+    private Car createCar(ResultSet resultSet) throws SQLException {
+        Long idCar = resultSet.getObject("id", long.class);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
        Long idCar = resultSet.getObject("id", Long.class);
```

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 35]
```diff
@@ -0,0 +1,210 @@
...
+        String modelCar = resultSet.getString("model");
+        Long manufacturerIdCar = resultSet.getObject("manufacturer_id", long.class);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: same everywhere - we need wrapper, not a primitive

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 36]
```diff
@@ -0,0 +1,210 @@
...
+        car.setId(idCar);
+        return car;
+    }
+
+    private List<Driver> getDriversForCar(Long idCar) {
+        String query = "SELECT drivers.name_driver, drivers.licence_number, "
+                + "cars_drivers.id_driver "
+                + "FROM cars_drivers cd "
+                + "JOIN drivers d ON cd.id_driver = d.id "
+                + "WHERE cd.id_car = ?";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]:
```suggestion
        String query = "SELECT d.name_driver, d.licence_number, "
                + "cd.id_driver "
                + "FROM cars_drivers cd "
                + "JOIN drivers d ON cd.id_driver = d.id "
                + "WHERE cd.id_car = ?"
                // and driver is not deleted;
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 37]
```diff
@@ -57,18 +57,19 @@ public Car create(Car car) {
             throw new DataProcessingException("Could not get car by id " + id, e);
         }
         if (car != null) {
-            getDriversForCar(car.getId());
+            car.setDrivers(getDriversForCar(car.getId()));
+
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: redundant empty line

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 38]
```diff
@@ -122,13 +121,13 @@ public boolean delete(Long id) {
 
     @Override
     public List<Car> getAllByDriver(Long driverId) {
-        String query = "SELECT cars.id, cars.model, cars.manufacturer_id, "
-                + "manufacturer.name, manufacturer.country "
+        String query = "SELECT c.id, c.model, c.manufacturer_id, "
+                + "m.name, m.country "
                 + "FROM cars c "
                 + "JOIN cars_drivers cd ON c.id = cd.id_car "
                 + "JOIN manufacturer m ON c.manufacturer_id = m.id "
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java:127]: we should join drivers as well and check that driver is not deleted

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 39]
```diff
@@ -184,8 +186,8 @@ private Car createCar(ResultSet resultSet) throws SQLException {
     }
 
     private List<Driver> getDriversForCar(Long idCar) {
-        String query = "SELECT drivers.name_driver, drivers.licence_number, "
-                + "cars_drivers.id_driver "
+        String query = "SELECT d.name_driver, d.licence_number, "
+                + "cd.id_driver "
                 + "FROM cars_drivers cd "
                 + "JOIN drivers d ON cd.id_driver = d.id "
                 + "WHERE cd.id_car = ?";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java]: we should check that driver is not deleted

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java [hunk 40]
```diff
@@ -190,7 +192,7 @@ private Car createCar(ResultSet resultSet) throws SQLException {
                 + "cd.id_driver "
                 + "FROM cars_drivers cd "
                 + "JOIN drivers d ON cd.id_driver = d.id "
-                + "WHERE cd.id_car = ?";
+                + "WHERE cd.id_car = ? AND d.beleted = FALSE";
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/dao/jdbc/CarDaoJdbc.java:194]: beleted -> deleted

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**